lib/crt.php

```diff
@@ -1,35 +1,93 @@
 <?php
-require_once('preload.php');
 
+// Wrapper for the FFI interface to the CRT. There only ever needs to be one of these.
 final class CRT {
     private static $ffi = null;
+    private static $refcount = 0;
+
     function __construct() {
         if (is_null(self::$ffi)) {
             try {
                 self::$ffi = FFI::cdef(file_get_contents(__DIR__ . "/api.h"), __DIR__ . "/libaws-crt-ffi.so");
+                self::init();
             } catch (Exception $e) {
-                echo 'Exception: ', $e->getMessage(), "\n";
+                echo 'Exception while initializing CRT via FFI: ', $e->getMessage(), "\n";
             }
         }
+        ++self::$refcount;
+    }
+
+    function __destruct() {
+        if (--self::$refcount == 0) {
+            self::clean_up();
+            self::$ffi = null;
+        }
     }
 
-    function init() {
+    // library lifetime
+    private static function init() {
+        //echo("Initializing CRT\n");
         return self::$ffi->aws_crt_init();
     }
 
-    function clean_up() {
+    private static function clean_up() {
+        //echo("Shutting down CRT\n");
         return self::$ffi->aws_crt_clean_up();
     }
 
-    function last_error() {
+    // Error handling
+    public static function last_error() {
         return self::$ffi->aws_crt_last_error();
     }
 
-    function error_str($error) {
+    public static function error_str($error) {
         return self::$ffi->aws_crt_error_str((int) $error);
     }
 
-    function error_name($error) {
+    public static function error_name($error) {
         return self::$ffi->aws_crt_error_name((int) $error);
     }
+
+    // IO
```

Comment:
Can you make this comment a little more descriptive? What you mean with IO here?